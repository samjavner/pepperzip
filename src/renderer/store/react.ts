import * as React from "react";
import { Commands, Dispatch, Store, Update, UseStoreOptions } from "./";

export interface Msg<M extends object, U extends Update<M>> {
    type: keyof U;
    payload?: any;
}

export function useLocalStore(log: (msg: any) => void = () => ({})) {
    return function actual<
        M extends object,
        U extends Update<M>,
        C extends Commands = {}
    >({
        init,
        update,
        createCommands,
        memo = [],
    }: UseStoreOptions<M, U, C>): Store<M, U, C> {
        const [model, reducerDispatch] = React.useReducer(
            (state: M, action: Msg<M, U>) =>
                update[action.type](state, action.payload),
            init
        );

        const vm = React.useMemo(
            () => {
                const dispatch: Dispatch<U> = {} as any;
                Object.keys(update).forEach(type => {
                    (dispatch as any)[type] = (payload: any) => {
                        const msg =
                            update[type].length === 1
                                ? { type }
                                : { type, payload };
                        log(msg);
                        reducerDispatch(msg);
                    };
                });
                const commands: C = createCommands
                    ? createCommands(model, dispatch)
                    : (({} as any) as C);
                return {
                    ...model,
                    ...dispatch,
                    ...commands,
                };
            },
            [model, update, ...memo]
        );

        return vm;
    };
}
