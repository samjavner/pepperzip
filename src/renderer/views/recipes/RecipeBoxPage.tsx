import { faPlus, faSearch } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import classNames from "classnames";
import * as React from "react";
import {
    ArrowKeyStepper,
    AutoSizer,
    defaultCellRangeRenderer,
    Grid,
    GridCellRangeProps,
} from "react-virtualized";
import * as uuid from "uuid";
import { RecipeBox } from "../../core/recipes/recipeBox";
import { Recipe } from "../../db/recipe";
import RecipeEditModal from "./RecipeEditModal";
import RecipeView from "./RecipeView";

/**
 * Implementation of cellRangeRenderer with caching disabled.
 * This is useful so that cells are re-rendered while stepping with arrow keys.
 */
function uncachedCellRangeRenderer(gridCellRangeProps: GridCellRangeProps) {
    return defaultCellRangeRenderer({
        ...gridCellRangeProps,
        cellCache: {},
    });
}

export const RecipeBoxPage: React.FunctionComponent<{
    recipeBox: RecipeBox;
}> = ({ recipeBox }) => {
    const recipe: Recipe | undefined =
        recipeBox.recipes[recipeBox.selectedIndex];

    return (
        <>
            <div
                style={{
                    height: "100%",
                    display: "grid",
                    gridTemplateColumns: "20rem 4fr",
                }}
            >
                <div
                    style={{
                        gridColumn: 1,
                        borderRight: "1px solid #c8cfd0", // $border
                        display: "grid",
                        gridTemplateRows: "auto 1fr",
                    }}
                >
                    <div
                        style={{
                            gridRow: 1,
                        }}
                    >
                        <div
                            className="field has-addons"
                            style={{ marginTop: "-1px" }}
                        >
                            <div className="control">
                                <input
                                    className="input"
                                    type="text"
                                    placeholder="Search recipes"
                                    style={{ width: "15rem", borderRadius: 0 }}
                                />
                            </div>
                            <div className="control">
                                <a
                                    className="button"
                                    href="#"
                                    style={{ borderRadius: 0 }}
                                >
                                    <FontAwesomeIcon icon={faSearch} />
                                </a>
                            </div>
                            <div className="control">
                                <a
                                    className="button"
                                    href="#"
                                    style={{ borderRadius: 0 }}
                                    onClick={recipeBox.openAddRecipe}
                                >
                                    <FontAwesomeIcon icon={faPlus} />
                                </a>
                            </div>
                        </div>
                    </div>
                    <div style={{ gridRow: 2, marginRight: "1px" }}>
                        <AutoSizer>
                            {({ height, width }) => (
                                <ArrowKeyStepper
                                    mode="cells"
                                    isControlled={true}
                                    columnCount={1}
                                    rowCount={recipeBox.recipes.length}
                                    scrollToColumn={1}
                                    scrollToRow={recipeBox.selectedIndex}
                                    onScrollToChange={({ scrollToRow }) =>
                                        recipeBox.setSelectedIndex(scrollToRow)
                                    }
                                >
                                    {({
                                        onSectionRendered,
                                        scrollToColumn,
                                        scrollToRow,
                                    }) => (
                                        <Grid
                                            onSectionRendered={
                                                onSectionRendered
                                            }
                                            scrollToColumn={scrollToColumn}
                                            scrollToRow={scrollToRow}
                                            height={height}
                                            width={width}
                                            rowCount={recipeBox.recipes.length}
                                            columnCount={1}
                                            rowHeight={25}
                                            columnWidth={width}
                                            autoContainerWidth={true}
                                            cellRangeRenderer={
                                                uncachedCellRangeRenderer
                                            }
                                            cellRenderer={({
                                                key,
                                                rowIndex,
                                                style,
                                            }) => (
                                                <div
                                                    key={key}
                                                    style={{
                                                        ...style,
                                                        overflow: "hidden",
                                                        textOverflow:
                                                            "ellipsis",
                                                        userSelect: "none",
                                                        whiteSpace: "nowrap",
                                                        width: "100%",
                                                    }}
                                                    className={classNames({
                                                        "has-background-link has-text-white is-selected-item":
                                                            rowIndex ===
                                                            recipeBox.selectedIndex,
                                                        "has-background-white":
                                                            rowIndex !==
                                                            recipeBox.selectedIndex,
                                                    })}
                                                    onClick={() =>
                                                        recipeBox.setSelectedIndex(
                                                            rowIndex
                                                        )
                                                    }
                                                >
                                                    {
                                                        recipeBox.recipes[
                                                            rowIndex
                                                        ].name
                                                    }
                                                </div>
                                            )}
                                            cell={true}
                                        />
                                    )}
                                </ArrowKeyStepper>
                            )}
                        </AutoSizer>
                    </div>
                </div>
                <div
                    style={{
                        gridColumn: 2,
                        overflowY: "scroll",
                        padding: "2rem",
                    }}
                >
                    {recipe && (
                        <RecipeView
                            recipe={recipe}
                            onEdit={recipeBox.openEditRecipe}
                        />
                    )}
                </div>
            </div>
            {recipeBox.isAddRecipeActive && (
                <RecipeEditModal
                    title="New Recipe"
                    init={{
                        id: uuid.v4(),
                        name: "New Recipe",
                        url: "",
                        description: "",
                        tags: [],
                        ingredients: "",
                        directions: "",
                        notes: "",
                        sourceText: "",
                        importWarnings: [],
                    }}
                    onSave={recipeBox.saveAddRecipe}
                    onClose={recipeBox.closeAddRecipe}
                />
            )}
            {recipeBox.isEditRecipeActive &&
                recipe && (
                    <RecipeEditModal
                        title="Edit Recipe"
                        init={recipe}
                        onSave={recipeBox.saveEditRecipe}
                        onClose={recipeBox.closeEditRecipe}
                    />
                )}
        </>
    );
};

export default RecipeBoxPage;
