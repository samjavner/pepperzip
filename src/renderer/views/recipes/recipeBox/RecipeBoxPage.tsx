import classNames from "classnames";
import * as React from "react";
import { connect } from "react-redux";
import {
    ArrowKeyStepper,
    AutoSizer,
    defaultCellRangeRenderer,
    Grid,
    GridCellRangeProps,
} from "react-virtualized";
import { Dispatch } from "redux";
import * as actions from "../../../app/actions";
import { State } from "../../../app/state";

function mapStateToProps(state: State) {
    return {
        selectedIndex: state.selectedRecipeIndex,
    };
}

function mapDispatchToProps(dispatch: Dispatch<actions.Action>) {
    return {
        setSelectedIndex: (index: number) =>
            dispatch(actions.setSelectedRecipeIndex(index)),
    };
}

/**
 * Implementation of cellRangeRenderer with caching disabled.
 * This is useful so that cells are re-rendered while stepping with arrow keys.
 */
function uncachedCellRangeRenderer(gridCellRangeProps: GridCellRangeProps) {
    return defaultCellRangeRenderer({
        ...gridCellRangeProps,
        cellCache: {},
    });
}

type RecipeBoxPageProps = ReturnType<typeof mapStateToProps> &
    ReturnType<typeof mapDispatchToProps>;

export const RecipeBoxPage: React.SFC<RecipeBoxPageProps> = props => (
    <div
        style={{
            height: "100%",
            display: "grid",
            gridTemplateColumns: "20rem 4fr",
        }}
    >
        <div style={{ gridColumn: 1 }}>
            <AutoSizer>
                {({ height, width }) => (
                    <ArrowKeyStepper
                        mode="cells"
                        isControlled={true}
                        columnCount={1}
                        rowCount={50000}
                        scrollToColumn={1}
                        scrollToRow={props.selectedIndex}
                        onScrollToChange={({ scrollToRow }) =>
                            props.setSelectedIndex(scrollToRow)
                        }
                    >
                        {({
                            onSectionRendered,
                            scrollToColumn,
                            scrollToRow,
                        }) => (
                            <Grid
                                onSectionRendered={onSectionRendered}
                                scrollToColumn={scrollToColumn}
                                scrollToRow={scrollToRow}
                                height={height}
                                width={width}
                                rowCount={50000}
                                columnCount={1}
                                rowHeight={25}
                                columnWidth={width}
                                autoContainerWidth={true}
                                cellRangeRenderer={uncachedCellRangeRenderer}
                                cellRenderer={({ key, rowIndex, style }) => (
                                    <div
                                        key={key}
                                        style={style}
                                        className={classNames({
                                            "has-background-link has-text-white":
                                                rowIndex ===
                                                props.selectedIndex,
                                        })}
                                        onClick={() =>
                                            props.setSelectedIndex(rowIndex)
                                        }
                                    >
                                        Recipe {rowIndex}
                                    </div>
                                )}
                                cell={true}
                            />
                        )}
                    </ArrowKeyStepper>
                )}
            </AutoSizer>
        </div>
        <div style={{ gridColumn: 2, overflowY: "auto" }}>
            <div>Index: {props.selectedIndex}</div>
            {[...Array(10).keys()].map(x => (
                <div className="section" key={x}>
                    Paragraph {x}
                </div>
            ))}
        </div>
    </div>
);

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(RecipeBoxPage);
