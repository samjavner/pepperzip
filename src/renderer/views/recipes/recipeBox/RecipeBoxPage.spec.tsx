import * as React from "react";
import * as renderer from "react-test-renderer";
import { RecipeBoxPage } from "./RecipeBoxPage";

describe("RecipeBoxPage", () => {
    it("renders the text 'Recipe Box'", () => {
        const component = renderer.create(<RecipeBoxPage />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
