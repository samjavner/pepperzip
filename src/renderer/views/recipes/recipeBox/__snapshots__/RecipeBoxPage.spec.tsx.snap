// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeBoxPage renders the master/detail view 1`] = `
<div
  style={
    Object {
      "display": "grid",
      "gridTemplateColumns": "1fr 4fr",
      "height": "100%",
    }
  }
>
  <div
    style={
      Object {
        "gridColumn": 1,
      }
    }
  >
    <div
      style={
        Object {
          "height": 0,
          "overflow": "visible",
          "width": 0,
        }
      }
    >
      <div
        onKeyDown={[Function]}
      >
        <div
          aria-label="grid"
          aria-readonly={true}
          className="ReactVirtualized__Grid"
          onScroll={[Function]}
          role="grid"
          style={
            Object {
              "WebkitOverflowScrolling": "touch",
              "boxSizing": "border-box",
              "direction": "ltr",
              "height": 0,
              "overflowX": "hidden",
              "overflowY": "auto",
              "position": "relative",
              "width": 0,
              "willChange": "transform",
            }
          }
          tabIndex={0}
        />
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "gridColumn": 2,
        "overflowY": "auto",
      }
    }
  >
    <div>
      Index: 
      10
    </div>
    <div
      className="section"
    >
      Paragraph 
      0
    </div>
    <div
      className="section"
    >
      Paragraph 
      1
    </div>
    <div
      className="section"
    >
      Paragraph 
      2
    </div>
    <div
      className="section"
    >
      Paragraph 
      3
    </div>
    <div
      className="section"
    >
      Paragraph 
      4
    </div>
    <div
      className="section"
    >
      Paragraph 
      5
    </div>
    <div
      className="section"
    >
      Paragraph 
      6
    </div>
    <div
      className="section"
    >
      Paragraph 
      7
    </div>
    <div
      className="section"
    >
      Paragraph 
      8
    </div>
    <div
      className="section"
    >
      Paragraph 
      9
    </div>
  </div>
</div>
`;
